signer/core/apitypes/types.go

```diff
@@ -490,13 +500,32 @@ func parseInteger(encType string, encValue interface{}) (*big.Int, error) {
 func (typedData *TypedData) EncodePrimitiveValue(encType string, encValue interface{}, depth int) ([]byte, error) {
 	switch encType {
 	case "address":
-		stringValue, ok := encValue.(string)
-		if !ok || !common.IsHexAddress(stringValue) {
-			return nil, dataMismatchError(encType, encValue)
+		stringValue, isString := encValue.(string)
+		if isString {
+			if !common.IsHexAddress(stringValue) {
+				return nil, dataMismatchError(encType, encValue)
+			}
+			retval := make([]byte, 32)
+			copy(retval[12:], common.HexToAddress(stringValue).Bytes())
+			return retval, nil
+		}
+		byteSlice, isByteSlice := encValue.([]byte)
+		if isByteSlice {
+			if len(byteSlice) == 20 {
+				retval := make([]byte, 32)
+				copy(retval[12:], byteSlice)
+				return retval, nil
+			} else {
+				return nil, errors.New("if a byte slice is passed in, in place of an address, it must be either 20 or 32 bytes long")
+			}
+		}
+		twentyByteArr, isTwentyByteArr := encValue.([20]byte)
+		if isTwentyByteArr {
+			retval := make([]byte, 32)
+			copy(retval[12:], twentyByteArr[:])
+			return retval, nil
```

Comment:
```suggestion
		retval := make([]byte, 32)
		switch val := encValue.(type) {
		case string:
			if common.IsHexAddress(val) {
				copy(retval[12:], common.HexToAddress(val).Bytes())
				return retval, nil
			}
		case []byte:
			if len(val) == 20 {
				copy(retval[12:], val)
				return retval, nil
			}
		case [20]byte:
			copy(retval[12:], val[:])
			return retval, nil
		}
```
IMO a bit clearer, and definitely shorter

Reply:
Also, we need tests for these type-conversions, (and preferrably the fail-cases too)

Reply:
Added tests for this, and adopted your change.